### PR TITLE
Implement XML to record converter

### PR DESCRIPTION
**Pull request by AzeemMuzammil**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/41180

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: distribution/zip/jballerina-tools/build.gradle [hunk 1]
```diff
@@ -136,6 +136,7 @@ dependencies {
     langserverLib project(':ls-extensions:partial-parser')
     langserverLib project(':ls-extensions:trigger-service')
     langserverLib project(':ls-extensions:bal-shell-service')
+    langserverLib project(':xml-to-record-converter')
```
Comment (NipunaRanasinghe) [inline distribution/zip/jballerina-tools/build.gradle:139]: any reason to deviate from the previous conventions? (we have added the json converter as `:ls-extensions:json-to-record-converter`)

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/Constants.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/*
+ *  Copyright (c) 2023, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/Constants.java:4]:
```suggestion
 *  Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org)
 *
 *  WSO2 LLC. licenses this file to you under the Apache License,
```

Comment (NipunaRanasinghe) [reply misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/Constants.java:4]: Apply the same to all the relevant files

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        NodeList<ModuleMemberDeclarationNode> moduleMembers =
+                AbstractNodeFactory.createNodeList(new ArrayList<>(typeDefNodes));;
+//        if (isRecordTypeDesc) {
+//            Optional<TypeDefinitionNode> lastTypeDefNode = convertToInlineRecord(typeDefNodes, diagnosticMessages);
+//            moduleMembers = lastTypeDefNode
+//                    .<NodeList<ModuleMemberDeclarationNode>>map(AbstractNodeFactory::createNodeList)
+//                    .orElseGet(AbstractNodeFactory::createEmptyNodeList);
+//        } else {
+//            moduleMembers = AbstractNodeFactory.createNodeList(new ArrayList<>(typeDefNodes));
+//        }
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:133]: any reason to keep this commented code?

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+//            DiagnosticMessage message = DiagnosticMessage.jsonToRecordConverter102(null);
+//            diagnosticMessages.add(message);
+        }
+//        if (!updatedFieldNames.isEmpty()) {
+//            updatedFieldNames.forEach((oldFieldName, updateFieldName) -> {
+//                DiagnosticMessage message =
+//                        DiagnosticMessage.jsonToRecordConverter106(new String[]{oldFieldName, updateFieldName});
+//                diagnosticMessages.add(message);
+//            });
+//        }
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:153]: same as above

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+            Element rootElement = doc.getDocumentElement();
+            generateRecords(rootElement, isClosed, recordToTypeDescNodes, diagnosticMessages);
+        } catch (ParserConfigurationException parserConfigurationException) {
+
+        } catch (SAXException saxException) {
+
+        } catch (IOException | IllegalArgumentException e) {
+
+        }
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:108]: Shall we handle these errors and return diagnostics, similar to what we are doing during the JSON to record conversion?

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordRequest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ * Request format for XMLToRecord endpoint.
+ *
+ * @since 2.0.0
+ */
+public class XMLToRecordRequest {
+
+    private String xmlValue;
+    private boolean isRecordTypeDesc;
+    private boolean isClosed;
+    private boolean forceFormatRecordFields;
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordRequest.java:31]: are all of these parameters applicable for the xml conversions as well?

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/util/ConverterUtils.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        return AbstractNodeFactory.createToken(SyntaxKind.STRING_KEYWORD);
+    }
+
+    private static boolean isBoolean(String value) {
+        return value.equalsIgnoreCase("true") || value.equalsIgnoreCase("false");
+    }
+
+    private static boolean isInteger(String value) {
+        try {
+            Integer.parseInt(value);
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/util/ConverterUtils.java:166]: IMO we should check if the value is compatible with java long instead (as Ballerina `int` is represented as `long` datatype in runtime)

```suggestion
            Long.parseLong(value);
```

**Review by NipunaRanasinghe — CHANGES_REQUESTED** [review]

**Review by NipunaRanasinghe — COMMENTED** [review]
@AzeemMuzammil as suggested by @daneshk, shall add the `@xmldata:Name` annotation only to the parent record? (having it in the child's record doesn't have any value as we map the XML element name with the field name)

File: distribution/zip/jballerina-tools/build.gradle (hunk 1 above)
Comment (AzeemMuzammil) [reply distribution/zip/jballerina-tools/build.gradle:139]: yeah, this module is not defines under ls-extensions dir,

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordRequest.java (hunk 6 above)
Comment (AzeemMuzammil) [reply misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordRequest.java:31]: yes, at the moment, not all are implemented, but yes

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/Constants.java (hunk 2 above)
Comment (rdulmina) [reply misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/Constants.java:4]: Seems It's LCC now not LLC

Comment (lochana-chathura) [reply misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/Constants.java:4]: Also, the url has to be ".com" not ".org"

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverterClientCapabilitySetter.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.xmltorecordconverter;
+
+
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverterClientCapabilitySetter.java:21]: Additional blank line

File: misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+            DiagnosticMessage message = DiagnosticMessage.xmlToRecordConverter101(null);
+            diagnosticMessages.add(message);
+            return DiagnosticUtils.getDiagnosticResponse(diagnosticMessages, response);
+        } catch (IOException | IllegalArgumentException e) {
+            DiagnosticMessage message = DiagnosticMessage.xmlToRecordConverter102(null);
+            diagnosticMessages.add(message);
+            return DiagnosticUtils.getDiagnosticResponse(diagnosticMessages, response);
+        }
+
+        io.ballerina.compiler.syntax.tree.NodeList<ImportDeclarationNode> imports =
```
Comment (NipunaRanasinghe) [inline misc/xml-to-record-converter/src/main/java/io/ballerina/xmltorecordconverter/XMLToRecordConverter.java:120]:
```suggestion
       NodeList<ImportDeclarationNode> imports =
```

**Review by NipunaRanasinghe — APPROVED** [review]
Approved with few minor suggestions

**Review by NipunaRanasinghe — APPROVED** [review]

**Pull request merged**